### ISSUE TITLE
Debug Adapter Protocol: Sending events after a response

### ISSUE THREAD

**Issue by fwcd** [eclipse/lsp4j#229]
The specification requires the debug adapter to send an `initialized` event back to the client *after* the `initialize` response has been returned:

![image](https://user-images.githubusercontent.com/30873659/44222978-8c84c080-a186-11e8-866e-e3c57f6c1076.png)

AFAIK, this is not possible using the `IDebugProtocolServer` interface without hacking deep into the JSON-RPC machinery. Is there a workaround for this?

**Comment by fwcd** [comment]
A thread is probably the easiest solution, but I would not want to cause any race conditions with clients that rely on the fact that `initialized` happens strictly after the capabilities response.

**Comment by jonahgraham** [comment]
You are correct that it is bad for the adapter to send the event back first, from the [spec](https://microsoft.github.io/debug-adapter-protocol/specification#Events_Initialized):

> A debug adapter is expected to send this [Initialized] event when it is ready to accept configuration requests (but not before the ‘initialize’ request has finished).

Frustratingly, for a long time the [Mock Debug Adapter](https://github.com/Microsoft/vscode-mock-debug/) for [VSCode did not respect this](https://github.com/Microsoft/vscode/issues/4902#issuecomment-363468536). As mock debug was the starting point for many implementations of debug adapters, The rest of that thread tries to clarify some of the details of how vscode operate, with a better picture of that is going on [here](https://github.com/Microsoft/vscode/issues/4902#issuecomment-368583522). As you can see VS Code sends requests concurrently to the DAP.

---

> A thread is probably the easiest solution, but I would not want to cause any race conditions with clients that rely on the fact that initialized happens strictly after the capabilities response.

I simply cannot see a way to do this with no possibility of a race condition based on current implementation of RemoteEndpoint. Strictly the problem isn't sending the initialized before the initialize response, but rather the problem is sending the launch response before the configuration done response because you can delay sending initialized until you receive the launch request.

This is what almost works, but it can theoretically run the thread defined in configurationDone() before configurationDone() returns, meaning that the launch reply would go before the configurationDone reply. For the current VSCode this is not a problem AFAICT.

```java
	// A thread pool executor
	ExecutorService executor;
	IDebugProtocolClient client;
	CompletableFuture<Void> launchFuture;

	@Override
	public CompletableFuture<Capabilities> initialize(InitializeRequestArguments args) {
		Capabilities capabilities = new Capabilities();
		capabilities.setSupportsConfigurationDoneRequest(true);
		// ... set other capabilities here
		return CompletableFuture.completedFuture(capabilities);
	}

	@Override
	public CompletableFuture<Void> launch(Map<String, Object> args) {
		client.initialized();

		// start debugger here, but don't complete it's launch until we get config done
		launchFuture = new CompletableFuture<>();
		return launchFuture;
	}

	@Override
	public CompletableFuture<SetBreakpointsResponse> setBreakpoints(SetBreakpointsArguments args) {
		// set breakpoints in debugger
		SetBreakpointsResponse setBreakpointsResponse = new SetBreakpointsResponse();
		// fill in details of setBreakpointsResponse
		return CompletableFuture.completedFuture(setBreakpointsResponse);
	}

	@Override
	public CompletableFuture<Void> configurationDone(ConfigurationDoneArguments args) {
		CompletableFuture<Void> configurationDone = new CompletableFuture<>();
		executor.execute(() -> {
			configurationDone.complete(null);
			launchFuture.complete(null);
		});
		return configurationDone;
	}
```

As a temporary (bad) workaround until we solve properly you can change configurationDone method to make sure that it has fully returned like this, as long as your executor service guarantees not to run runnable within the calling thread!

```
	@Override
	public CompletableFuture<Void> configurationDone(ConfigurationDoneArguments args) {
		CompletableFuture<Void> configurationDone = new CompletableFuture<>();
		executor.execute(() -> {
			while (configurationDone.getNumberOfDependents() == 0) {
				// wait to make sure return of configurationDone() method is done and
				// the dependent has been added before marking configurationDone as complete
				// to ensure in order operation.
			}
			configurationDone.complete(null);
			launchFuture.complete(null);
		});
		return configurationDone;
	}
```

---

So, now on to the (proper) solution. It is possible to customize DebugRemoteEndpoint to allow you to hook into after RemoteEndpoint.handleRequest(RequestMessage) runs to ensure order of operations. I don't know how this looks at the moment, so I will get back to you on this.

**Comment by jonahgraham** [comment]
PS. In VSCode, but not in the spec, what happens when client receives configurationDone is to send a threads request. So the other temporary workaround is to complete launchFuture at the top of threads request.

**Comment by FeldrinH** [comment]
Has anything changed? Is there now a cleaner solution to the problem integrated into lsp4j?

**Comment by FeldrinH** [comment]
It would be really nice if there was built-in support for this in lsp4j.

As far as I can tell the suggestion of hooking into `RemoteEndpoint.handleRequest(RequestMessage)` can only be implemented inside lsp4j, because `handleRequest` does not expose the future that you need to hook and replacing the `handleRequest` implementation in a subclass of `RemoteEndpoint` can't be done without a lot of dirty hacks, because it uses a private field in `RemoteEndpoint` (`receivedRequestMap`).

**Comment by jonahgraham** [comment]
> Has anything changed? Is there now a cleaner solution to this problem integrated into lsp4j?

Nothing has changed - but from your recent comment you seem to have some concrete ideas on how to improve it. Can you turn that into a PR?

**Comment by FeldrinH** [comment]
I have a usable but somewhat dirty workaround that hooks into `MessageConsumer.consume` instead implemented for https://github.com/goblint/GobPie/pull/60. I have a workable idea for a clean technical implementation, but I don't have a good idea for the public API. I'll look into turning it into a PR when I have time (possibly sometime next week).
